android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentService.java

```diff
@@ -0,0 +1,876 @@
+        @Override
+        public void run() {
+            while (true) {
+                try {
+                    Thread.sleep(OBSERVER_PING_PERIOD_MS);
+                    Set<Messenger> lifecycleObserversCopy = ConcurrentHashMap.newKeySet();
+                    synchronized (lifecycleObservers) {
+                        lifecycleObserversCopy.addAll(lifecycleObservers);
+                    }
+                    int aliveObservers = 0;
+                    for (Iterator<Messenger> it = lifecycleObserversCopy.iterator(); it.hasNext(); ) {
+                        Messenger messenger = it.next();
+                        try {
+                            Message pingMsg = Message.obtain(null, LIFECYCLE_MSG_PING_REQUEST);
+                            Bundle pingData = new Bundle();
+                            UUID pingUuid = UUID.randomUUID();
+                            pingData.putString(EXTRA_OBSERVER_PING_DATA, pingUuid.toString());
+                            CompletableFuture<UUID> pongReplyFuture = new CompletableFuture<>();
+                            synchronized (pingRequests) {
+                                pingRequests.put(pingUuid, pongReplyFuture);
+                            }
+                            pingMsg.setData(pingData);
+                            messenger.send(pingMsg);
+
+                            try {
+                                pongReplyFuture.get(OBSERVER_PING_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+                                aliveObservers++;
+                            } catch (ExecutionException | TimeoutException e) {
+                                Log.w(LOG_TAG, "No pong received. " + e);
+                                synchronized (pingRequests) {
+                                    pingRequests.remove(pingUuid);
+                                }
+                            } catch (InterruptedException ignore) {
+                                return;
+                            }
+                        } catch (RemoteException | IllegalStateException e) {
+                            // probably an observer has crashed or terminated
+                            Log.w(LOG_TAG, "Unable to send ping request " + Log.getStackTraceString(e));
+                        }
+
+                        if (0 == aliveObservers) {
+                            onNoObserversLeft();
+                        }
+                    }
+                } catch (InterruptedException ignore) {
+                    Log.d(LOG_TAG, "Ping thread interrupted");
+                }
+            }
+        }
+    }
+
+    /**
+     * Stops sending periodic ping messages to the lifecycle observers.
+     */
+    protected final void stopPingThread() {
+        PingThread thread = pingThread.getAndSet(null);
+        if (thread != null && thread.isAlive()) {
+            thread.interrupt();
+            try {
+                thread.join();
+            } catch (InterruptedException ignore) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)